Source/WebCore/Modules/url-pattern/URLPatternParser.cpp

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternParser.h"
+
+#include "URLPatternTokenizer.h"
+
+namespace WebCore {
+namespace URLPatternUtilities {
+
+URLPatternParser::URLPatternParser(EncodingCallbackType type, String&& segmentWildcardRegexp)
+    : m_callbackType(type)
+    , m_segmentWildcardRegexp(WTFMove(segmentWildcardRegexp))
+{
+}
+
+ExceptionOr<void> URLPatternParser::performParse(const URLPatternStringOptions& patternSettings)
+{
+    ExceptionOr<void> maybeFunctionException;
+
+    while (m_index < m_tokenList.size()) {
+        auto charToken = tryToConsumeToken(TokenType::Char);
+        auto nameToken = tryToConsumeToken(TokenType::Name);
+        auto regexOrWildcardToken = tryToConsumeRegexOrWildcardToken(nameToken);
+
+        if (!nameToken.isNull() || !regexOrWildcardToken.isNull()) {
+            String prefix;
+
+            if (!charToken.isNull())
+                prefix = charToken.value.toString();
+
+            if (!prefix.isEmpty() && prefix != patternSettings.prefixCodepoint)
+                std::exchange(m_pendingFixedValue, makeString(m_pendingFixedValue, std::exchange(prefix, { })));
```

Comment:
Spec says sets prefix to empty string, not null.
Also it's more clear to write it as:
`m_pendingFixedValue = makeString(m_pendingFixedValue, std::exchange(prefix, emptyString());`

If we expect to keep appending to `m_pendingFixedValue`, it should probably be a `StringBuilder`.